File: test-touch-targets.js
#!/usr/bin/env node

/**
 * Touch Target Verification Test
 * Tests touch target sizes across mobile viewports (320px-768px)
 * WCAG 2.5.5 Level AAA: Minimum 44x44px
 * Google Material Design: Minimum 48x48px recommended for mobile
 */

const fs = require('fs');
const path = require('path');

// WCAG 2.5.5 Level AAA minimum
const WCAG_MIN_SIZE = 44;
// Google Material Design recommendation
const GOOGLE_MIN_SIZE = 48;
// Recommended spacing between touch targets
const MIN_SPACING = 8;

// Test viewports
const VIEWPORTS = [
    { name: 'iPhone SE', width: 320, minSize: GOOGLE_MIN_SIZE },
    { name: 'Mobile Portrait', width: 480, minSize: GOOGLE_MIN_SIZE },
    { name: 'Tablet Portrait', width: 768, minSize: WCAG_MIN_SIZE }
];

// Interactive elements that should have proper touch targets
const TOUCH_TARGETS = [
    { selector: '#audio-controls button', name: 'Audio Control Buttons', context: 'all' },
    { selector: '.volume-slider input[type="range"]', name: 'Volume Sliders', context: 'all' },
    { selector: '.difficulty-btn', name: 'Difficulty Buttons', context: 'all' },
    { selector: '#start-button', name: 'Start Button', context: 'all' },
    { selector: '#help-button', name: 'Help Button', context: 'all' },
    { selector: '#pause-button', name: 'Pause Button', context: 'all' },
    { selector: '#help-close', name: 'Help Close Button', context: 'all' },
    { selector: '#pause-overlay button', name: 'Pause Overlay Buttons', context: 'all' },
    { selector: '#resume-button', name: 'Resume Button', context: 'all' },
    { selector: '#restart-button', name: 'Restart Button', context: 'all' }
];

class TouchTargetAnalyzer {
    constructor() {
        this.results = {
            timestamp: new Date().toISOString(),
            viewports: {},
            summary: {
                totalTests: 0,
                passed: 0,
                failed: 0,
                warnings: 0
            }
        };
    }

    parseCSS(htmlContent) {
        // Extract CSS rules from HTML
        const styleMatch = htmlContent.match(/<style>([\s\S]*?)<\/style>/);
        if (!styleMatch) return {};

        const cssContent = styleMatch[1];
        const rules = {};

        // Parse basic CSS rules (simplified parser)
        const ruleRegex = /([^{]+)\{([^}]+)\}/g;
        let match;

        while ((match = ruleRegex.exec(cssContent)) !== null) {
            const selector = match[1].trim();
            const declarations = match[2].trim();

            const properties = {};
            declarations.split(';').forEach(decl => {
                const [prop, value] = decl.split(':').map(s => s.trim());
                if (prop && value) {
                    properties[prop] = value;
                }
            });

            rules[selector] = properties;
        }

        return rules;
    }

    getEffectiveStyle(selector, cssRules, viewport) {
        const style = {
            minWidth: null,
            minHeight: null,
            width: null,
            height: null,
            padding: { top: 0, right: 0, bottom: 0, left: 0 }
        };

        // Check base styles
        if (cssRules[selector]) {
            this.applyStyles(style, cssRules[selector]);
        }

        // Check media queries for mobile
        if (viewport.width <= 768) {
            const mediaQuery768 = `@media (max-width: 768px)`;
            Object.keys(cssRules).forEach(rule => {
                if (rule.includes(mediaQuery768)) {
                    const innerRules = this.parseMediaQuery(rule, cssRules);
                    if (innerRules[selector]) {
                        this.applyStyles(style, innerRules[selector]);
                    }
                }
            });
        }

        if (viewport.width <= 480) {
            const mediaQuery480 = `@media (max-width: 480px)`;
            Object.keys(cssRules).forEach(rule => {
                if (rule.includes(mediaQuery480)) {
                    const innerRules = this.parseMediaQuery(rule, cssRules);
                    if (innerRules[selector]) {
                        this.applyStyles(style, innerRules[selector]);
                    }
                }
            });
        }

        return style;
    }

    applyStyles(style, properties) {
        if (properties['min-width']) {
            style.minWidth = this.parseSize(properties['min-width']);
        }
        if (properties['min-height']) {
            style.minHeight = this.parseSize(properties['min-height']);
        }
        if (properties['width']) {
            style.width = this.parseSize(properties['width']);
        }
        if (properties['height']) {
            style.height = this.parseSize(properties['height']);
        }
        if (properties['padding']) {
            const padding = this.parsePadding(properties['padding']);
            if (padding) {
                style.padding = padding;
            }
        }
    }

    parseSize(value) {
        if (!value) return null;
        const match = value.match(/(\d+)px/);
        return match ? parseInt(match[1]) : null;
    }

    parsePadding(value) {
        if (!value) return null;
        const parts = value.split(' ').map(v => this.parseSize(v)).filter(v => v !== null);
        if (parts.length === 1) {
            return { top: parts[0], right: parts[0], bottom: parts[0], left: parts[0] };
        } else if (parts.length === 2) {
            return { top: parts[0], right: parts[1], bottom: parts[0], left: parts[1] };
        } else if (parts.length === 4) {
            return { top: parts[0], right: parts[1], bottom: parts[2], left: parts[3] };
        }
        return { top: 0, right: 0, bottom: 0, left: 0 };
    }

    calculateTouchTarget(style) {
        // Calculate effective touch target size including padding
        let width = style.minWidth || style.width || 0;
        let height = style.minHeight || style.height || 0;

        // Add padding to dimensions
        width += style.padding.left + style.padding.right;
        height += style.padding.top + style.padding.bottom;

        return { width, height };
    }

    analyzeTarget(target, cssRules, viewport) {
        const test = {
            selector: target.selector,
            name: target.name,
            viewport: viewport.name,
            expected: viewport.minSize,
            actual: { width: 0, height: 0 },
            status: 'unknown',
            issues: []
        };

        // Get computed styles for this selector
        const style = this.getEffectiveStyle(target.selector, cssRules, viewport);
        const touchTarget = this.calculateTouchTarget(style);

        test.actual = touchTarget;

        // Check if meets minimum requirements
        const meetsWidth = touchTarget.width >= viewport.minSize;
        const meetsHeight = touchTarget.height >= viewport.minSize;

        if (meetsWidth && meetsHeight) {
            test.status = 'pass';
            this.results.summary.passed++;
        } else {
            test.status = 'fail';
            this.results.summary.failed++;

            if (!meetsWidth) {
                test.issues.push(`Width ${touchTarget.width}px is less than minimum ${viewport.minSize}px`);
            }
            if (!meetsHeight) {
                test.issues.push(`Height ${touchTarget.height}px is less than minimum ${viewport.minSize}px`);
            }
        }

        // Additional warnings
        if (touchTarget.width < GOOGLE_MIN_SIZE || touchTarget.height < GOOGLE_MIN_SIZE) {
            test.issues.push(`Below Google's recommended 48x48px for mobile`);
            this.results.summary.warnings++;
        }

        this.results.summary.totalTests++;
        return test;
    }

    async analyze(htmlPath) {
        console.log('🔍 Touch Target Verification Test');
        console.log('=================================\n');

        const htmlContent = fs.readFileSync(htmlPath, 'utf-8');
        const cssRules = this.parseCSS(htmlContent);

        // Test each viewport
        for (const viewport of VIEWPORTS) {
            console.log(`📱 Testing ${viewport.name} (${viewport.width}px)`);
            console.log(`   Minimum requirement: ${viewport.minSize}x${viewport.minSize}px\n`);

            this.results.viewports[viewport.name] = {
                width: viewport.width,
                minSize: viewport.minSize,
                tests: []
            };

            for (const target of TOUCH_TARGETS) {
                const result = this.analyzeTarget(target, cssRules, viewport);
                this.results.viewports[viewport.name].tests.push(result);

                const icon = result.status === 'pass' ? '✅' : '❌';
                console.log(`   ${icon} ${result.name} (${result.selector})`);
                console.log(`      Size: ${result.actual.width}x${result.actual.height}px`);

                if (result.issues.length > 0) {
                    result.issues.forEach(issue => {
                        console.log(`      ⚠️  ${issue}`);
                    });
                }
                console.log('');
            }
        }

        return this.results;
    }

    generateReport() {
        const { totalTests, passed, failed, warnings } = this.results.summary;
        const passRate = ((passed / totalTests) * 100).toFixed(1);

        console.log('\n📊 Test Summary');
        console.log('================');
        console.log(`Total Tests: ${totalTests}`);
        console.log(`✅ Passed: ${passed}`);
        console.log(`❌ Failed: ${failed}`);
        console.log(`⚠️  Warnings: ${warnings}`);
        console.log(`Pass Rate: ${passRate}%\n`);

        if (failed === 0) {
            console.log('🎉 All touch targets meet accessibility requirements!');
        } else {
            console.log('⚠️  Some touch targets need adjustment to meet requirements.');
        }

        return this.results;
    }

    saveReport(outputPath) {
        fs.writeFileSync(outputPath, JSON.stringify(this.results, null, 2));
        console.log(`\n💾 Detailed report saved to: ${outputPath}`);
    }
}

// Run the analyzer
async function main() {
    const analyzer = new TouchTargetAnalyzer();
    const htmlPath = path.join(__dirname, 'index.html');
    const outputPath = path.join(__dirname, 'touch-target-test-results.json');

    try {
        await analyzer.analyze(htmlPath);
        analyzer.generateReport();
        analyzer.saveReport(outputPath);

        // Exit with appropriate code
        process.exit(analyzer.results.summary.failed === 0 ? 0 : 1);
    } catch (error) {
        console.error('❌ Test failed with error:', error);
        process.exit(1);
    }
}

if (require.main === module) {
    main();
}

module.exports = { TouchTargetAnalyzer, VIEWPORTS, TOUCH_TARGETS };

File: test-touch-targets.py
#!/usr/bin/env python3

"""
Touch Target Verification Test
Tests touch target sizes across mobile viewports (320px-768px)
WCAG 2.5.5 Level AAA: Minimum 44x44px
Google Material Design: Minimum 48x48px recommended for mobile
"""

import re
import json
from datetime import datetime
from pathlib import Path

# WCAG 2.5.5 Level AAA minimum
WCAG_MIN_SIZE = 44
# Google Material Design recommendation
GOOGLE_MIN_SIZE = 48
# Recommended spacing between touch targets
MIN_SPACING = 8

# Test viewports
VIEWPORTS = [
    {'name': 'iPhone SE', 'width': 320, 'minSize': GOOGLE_MIN_SIZE},
    {'name': 'Mobile Portrait', 'width': 480, 'minSize': GOOGLE_MIN_SIZE},
    {'name': 'Tablet Portrait', 'width': 768, 'minSize': WCAG_MIN_SIZE}
]

# Interactive elements that should have proper touch targets
TOUCH_TARGETS = [
    {'selector': '#audio-controls button', 'name': 'Audio Control Buttons', 'context': 'all'},
    {'selector': '.volume-slider input[type="range"]', 'name': 'Volume Sliders', 'context': 'all'},
    {'selector': '.difficulty-btn', 'name': 'Difficulty Buttons', 'context': 'all'},
    {'selector': '#start-button', 'name': 'Start Button', 'context': 'all'},
    {'selector': '#help-button', 'name': 'Help Button', 'context': 'all'},
    {'selector': '#pause-button', 'name': 'Pause Button', 'context': 'all'},
    {'selector': '#help-close', 'name': 'Help Close Button', 'context': 'all'},
    {'selector': '#pause-overlay button', 'name': 'Pause Overlay Buttons', 'context': 'all'},
    {'selector': '#resume-button', 'name': 'Resume Button', 'context': 'all'},
    {'selector': '#restart-button', 'name': 'Restart Button', 'context': 'all'}
]

class TouchTargetAnalyzer:
    def __init__(self):
        self.results = {
            'timestamp': datetime.now().isoformat(),
            'viewports': {},
            'summary': {
                'totalTests': 0,
                'passed': 0,
                'failed': 0,
                'warnings': 0
            }
        }
        self.css_rules = {}

    def parse_css(self, html_content):
        """Extract and parse CSS rules from HTML"""
        style_match = re.search(r'<style>(.*?)</style>', html_content, re.DOTALL)
        if not style_match:
            return {}

        css_content = style_match.group(1)

        # Parse media queries separately
        media_queries = {}
        media_pattern = r'@media\s*\([^)]+\)\s*\{([^}]*(?:\{[^}]*\}[^}]*)*)\}'

        for match in re.finditer(media_pattern, css_content, re.DOTALL):
            media_query = match.group(0)
            media_content = match.group(1)

            # Extract max-width from media query
            max_width_match = re.search(r'max-width:\s*(\d+)px', media_query)
            if max_width_match:
                max_width = int(max_width_match.group(1))
                media_queries[max_width] = self.parse_rules(media_content)

        # Parse base rules (outside media queries)
        base_css = re.sub(media_pattern, '', css_content, flags=re.DOTALL)
        base_rules = self.parse_rules(base_css)

        return {'base': base_rules, 'media': media_queries}

    def parse_rules(self, css_content):
        """Parse CSS rules"""
        rules = {}
        rule_pattern = r'([^{]+)\{([^}]+)\}'

        for match in re.finditer(rule_pattern, css_content):
            selector = match.group(1).strip()
            declarations = match.group(2).strip()

            properties = {}
            for decl in declarations.split(';'):
                if ':' in decl:
                    prop, value = decl.split(':', 1)
                    properties[prop.strip()] = value.strip()

            rules[selector] = properties

        return rules

    def get_effective_style(self, selector, viewport):
        """Get effective style for a selector at a given viewport"""
        style = {
            'min-width': None,
            'min-height': None,
            'width': None,
            'height': None,
            'padding': [0, 0, 0, 0]  # top, right, bottom, left
        }

        # Apply base styles
        base_rules = self.css_rules.get('base', {})
        if selector in base_rules:
            self.apply_styles(style, base_rules[selector])

        # Apply media query styles (in order of specificity)
        media_queries = self.css_rules.get('media', {})
        for max_width in sorted(media_queries.keys(), reverse=True):
            if viewport['width'] <= max_width:
                media_rules = media_queries[max_width]
                if selector in media_rules:
                    self.apply_styles(style, media_rules[selector])

        return style

    def apply_styles(self, style, properties):
        """Apply CSS properties to style dict"""
        if 'min-width' in properties:
            style['min-width'] = self.parse_size(properties['min-width'])
        if 'min-height' in properties:
            style['min-height'] = self.parse_size(properties['min-height'])
        if 'width' in properties:
            style['width'] = self.parse_size(properties['width'])
        if 'height' in properties:
            style['height'] = self.parse_size(properties['height'])
        if 'padding' in properties:
            padding = self.parse_padding(properties['padding'])
            if padding:
                style['padding'] = padding

    def parse_size(self, value):
        """Parse a CSS size value (px only)"""
        if not value:
            return None
        match = re.search(r'(\d+)px', value)
        return int(match.group(1)) if match else None

    def parse_padding(self, value):
        """Parse CSS padding value"""
        if not value:
            return None

        parts = []
        for part in value.split():
            size = self.parse_size(part)
            if size is not None:
                parts.append(size)

        if len(parts) == 1:
            return [parts[0], parts[0], parts[0], parts[0]]
        elif len(parts) == 2:
            return [parts[0], parts[1], parts[0], parts[1]]
        elif len(parts) == 4:
            return parts

        return [0, 0, 0, 0]

    def calculate_touch_target(self, style):
        """Calculate effective touch target size including padding"""
        width = style['min-width'] or style['width'] or 0
        height = style['min-height'] or style['height'] or 0

        # Add padding to dimensions
        padding = style['padding']
        width += padding[1] + padding[3]  # right + left
        height += padding[0] + padding[2]  # top + bottom

        return {'width': width, 'height': height}

    def analyze_target(self, target, viewport):
        """Analyze a single touch target"""
        test = {
            'selector': target['selector'],
            'name': target['name'],
            'viewport': viewport['name'],
            'expected': viewport['minSize'],
            'actual': {'width': 0, 'height': 0},
            'status': 'unknown',
            'issues': []
        }

        # Get computed styles for this selector
        style = self.get_effective_style(target['selector'], viewport)
        touch_target = self.calculate_touch_target(style)

        test['actual'] = touch_target

        # Check if meets minimum requirements
        meets_width = touch_target['width'] >= viewport['minSize']
        meets_height = touch_target['height'] >= viewport['minSize']

        if meets_width and meets_height:
            test['status'] = 'pass'
            self.results['summary']['passed'] += 1
        else:
            test['status'] = 'fail'
            self.results['summary']['failed'] += 1

            if not meets_width:
                test['issues'].append(
                    f"Width {touch_target['width']}px is less than minimum {viewport['minSize']}px"
                )
            if not meets_height:
                test['issues'].append(
                    f"Height {touch_target['height']}px is less than minimum {viewport['minSize']}px"
                )

        # Additional warnings
        if touch_target['width'] < GOOGLE_MIN_SIZE or touch_target['height'] < GOOGLE_MIN_SIZE:
            if touch_target['width'] >= WCAG_MIN_SIZE and touch_target['height'] >= WCAG_MIN_SIZE:
                test['issues'].append(
                    "Below Google's recommended 48x48px for mobile (but meets WCAG 44x44px)"
                )
                self.results['summary']['warnings'] += 1

        self.results['summary']['totalTests'] += 1
        return test

    def analyze(self, html_path):
        """Analyze touch targets in HTML file"""
        print('🔍 Touch Target Verification Test')
        print('=================================\n')

        with open(html_path, 'r') as f:
            html_content = f.read()

        self.css_rules = self.parse_css(html_content)

        # Test each viewport
        for viewport in VIEWPORTS:
            print(f"📱 Testing {viewport['name']} ({viewport['width']}px)")
            print(f"   Minimum requirement: {viewport['minSize']}x{viewport['minSize']}px\n")

            self.results['viewports'][viewport['name']] = {
                'width': viewport['width'],
                'minSize': viewport['minSize'],
                'tests': []
            }

            for target in TOUCH_TARGETS:
                result = self.analyze_target(target, viewport)
                self.results['viewports'][viewport['name']]['tests'].append(result)

                icon = '✅' if result['status'] == 'pass' else '❌'
                print(f"   {icon} {result['name']} ({result['selector']})")
                print(f"      Size: {result['actual']['width']}x{result['actual']['height']}px")

                if result['issues']:
                    for issue in result['issues']:
                        print(f"      ⚠️  {issue}")
                print()

        return self.results

    def generate_report(self):
        """Generate summary report"""
        summary = self.results['summary']
        total_tests = summary['totalTests']
        passed = summary['passed']
        failed = summary['failed']
        warnings = summary['warnings']
        pass_rate = (passed / total_tests * 100) if total_tests > 0 else 0

        print('\n📊 Test Summary')
        print('================')
        print(f"Total Tests: {total_tests}")
        print(f"✅ Passed: {passed}")
        print(f"❌ Failed: {failed}")
        print(f"⚠️  Warnings: {warnings}")
        print(f"Pass Rate: {pass_rate:.1f}%\n")

        if failed == 0:
            print('🎉 All touch targets meet accessibility requirements!')
        else:
            print('⚠️  Some touch targets need adjustment to meet requirements.')

        return self.results

    def save_report(self, output_path):
        """Save detailed report to JSON file"""
        with open(output_path, 'w') as f:
            json.dump(self.results, f, indent=2)
        print(f"\n💾 Detailed report saved to: {output_path}")


def main():
    analyzer = TouchTargetAnalyzer()
    html_path = Path(__file__).parent / 'index.html'
    output_path = Path(__file__).parent / 'touch-target-test-results.json'

    try:
        analyzer.analyze(html_path)
        analyzer.generate_report()
        analyzer.save_report(output_path)

        # Exit with appropriate code
        exit(0 if analyzer.results['summary']['failed'] == 0 else 1)
    except Exception as error:
        print(f'❌ Test failed with error: {error}')
        import traceback
        traceback.print_exc()
        exit(1)


if __name__ == '__main__':
    main()
